File: TradingEAC-Java-StockService/src/main/java/tradingeac/java/stockservice/agents/SymbolAgent.java
package tradingeac.java.stockservice.agents;

import java.util.ArrayList;
import java.util.List;

import org.springframework.context.annotation.Configuration;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.web.reactive.function.client.WebClient;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import tradingeac.java.stockservice.models.StockSymbol;
import tradingeac.java.stockservice.util.cache.GenericCache;

@Configuration
public class SymbolAgent {
    private final String SymbolUrl = "https://finnhub.io/api/v1/stock/symbol";
    private final String Token = "[TOKEN_FINHUB]";

    private final WebClient _client;    
    private final GenericCache<String, List<StockSymbol>> _cache;
            
    public SymbolAgent(WebClient.Builder webClientBuilder) {
        _client = webClientBuilder.baseUrl(SymbolUrl).build(); 
        _cache = new GenericCache<String, List<StockSymbol>>(360000L);
    }
    
    public Flux<StockSymbol> listStockSymbols()
    {
        String key = "ListStockSymbols";        
        List<StockSymbol> stockSymbols = _cache.get(key).orElse(null);
        ParameterizedTypeReference<List<StockSymbol>> typeReference =  new ParameterizedTypeReference<List<StockSymbol>>() {};

        if(stockSymbols != null){
            return Mono.just(stockSymbols).flatMapMany(Flux::fromIterable);            
        }else{
            return _client.get()
            .uri("?exchange=US&token={Token}",Token)
            .retrieve()            
            .bodyToMono(typeReference)
            .map(sb->{
                _cache.put(key, sb);
                return sb;
            })
            .flatMapMany(Flux::fromIterable);
        }
    }
}

File: TradingEAC-DotNet-StockService/Agents/SymbolAgent.cs
using Microsoft.Extensions.Caching.Memory;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using TradingEAC.DotNet.StockService.Models;

namespace TradingEAC.DotNet.StockService.Agents
{
    public class SymbolAgent
    {
        private readonly HttpClient _client;
        private IMemoryCache _cache;

        //https://finnhub.io/api/v1/stock/symbol?exchange=US&token=[TOKEN_FINHUB]"
        private readonly string SymbolUrl = "https://finnhub.io/api/v1/stock/symbol";
        private readonly string Token = "[TOKEN_FINHUB]";
        public SymbolAgent(HttpClient client, IMemoryCache cache)
        {
            client.BaseAddress = new Uri(SymbolUrl);            
            _client = client;
            _cache = cache;
        }

        public async Task<IEnumerable<StockSymbol>> ListStockSymbols()
        {
            IEnumerable<StockSymbol> stockSymbols;
            string key = "ListStockSymbols";

            if (!_cache.TryGetValue(key, out stockSymbols))
            {
                var response = await _client.GetAsync($"?exchange=US&token={Token}");
                response.EnsureSuccessStatusCode();
                using var responseStream = await response.Content.ReadAsStreamAsync();
                stockSymbols = await JsonSerializer.DeserializeAsync<IEnumerable<StockSymbol>>(responseStream);

                var cacheEntryOptions = new MemoryCacheEntryOptions()
                   .SetSlidingExpiration(TimeSpan.FromMinutes(60));

                _cache.Set(key, stockSymbols, cacheEntryOptions);                
            }

            return stockSymbols;
        }
    }
}
